### PR TITLE
flash: add support for nvram devices

### PR DESCRIPTION
**Pull request by Laczen**
Flash devices in zephyr can be divided into devices that impose write restrictions (i.e. a write requires an area to be erased first) and devices that do not impose write restrictions (i.e. write can be performed at all times).

Examples of devices that do not impose write restrictions: ambiq, numaker_rmc, nrf_rram, nrf_mram and at45 in `read-modify-write` mode.

This PR adds support for being able to distinguish between those 2 types of devices and adds a boolean to the `flash_parameters` to be able to distinguish between them. The PR is an alternative to #67687.

The starting point of this PR is that no change should be made to any existing systems that cannot take advantage of the existence of the unrestricted write possibilities that is available. It should also not change the current understanding of the
flash api. The PR does not try to move the flash api into a more generic storage api.

*The rationale behind this starting point is to provide continuity for developers that need to provide support for multiple versions of zephyr, they should not have to figure out the meaning of the flash api in different versions of zephyr.*

The PR extends the `flash_parameters` with a bool `unrestricted_write` to allows subsystems to take advantage of this. The PR provides 2 configuration options `CONFIG_FLASH_ANY_HAS_RESTRICTED_WRITE` and `CONFIG_FLASH_ANY_HAS_UNRESTRICTED_WRITE` that can be used for code optimization.

The `flash_simulator` is extended with an option to simulate devices that have the unrestricted write property.

The `flash_disk` driver is used as an example of how a subsystem/driver can take advantage of the unrestricted write property.

### CODE REVIEW

File: drivers/flash/Kconfig
Comment (nordicjm) [inline drivers/flash/Kconfig:36]: This is a hardware feature, why is it being described in Kconfig? And can you explain how this would work with SFDP?

Comment (Laczen) [reply drivers/flash/Kconfig:36]: > This is a hardware feature, why is it being described in Kconfig? 

At the moment this feature is considered to be part of the `compatible` and the property (which is reflecting what kind of devices are available, hence the `_ANY_`) is being set by the driver Kconfig. It could be reflected in dts but this would more-or-less make it configurable. If it is deemed required that this should also be configurable it can be added as a devicetree property. This might require some macrobatics to set the property to the correct value, but should be doable.

> And can you explain how this would work with SFDP?

I suppose you are referring to the fact that it might be needed to discover the property from SFDP. This could be done as:
1. If a compatible needs to support both `restricted_write` and `unrestricted_write` the driver Kconfig enables both properties, the driver bool is then used in subsystems that can take advantage of it. This of course would eliminate any possible code optimizations as it is considered not known at compile time what the properties are. For systems that are not taking advantage of it nothing changes as they are not using the driver bool.
2. To enable back the optimization a property could be added to the devicetree description that forces a certain operation mode. This devicetree property can than be used in the Kconfig to only enable the expected operation mode. To avoid runtime errors a runtime check should be made to make sure that the forced mode is supported by the hardware.

This is only one way, other methods that would "split up" the driver into several device types would also be possible.

**Review by de-nordic — CHANGES_REQUESTED** [review]
This PR is not an alternative for https://github.com/zephyrproject-rtos/zephyr/pull/67687 it freeloads on that PR and discussion around it, yet it completely fails to address base issues that the original PR tries to solve.
Ideas of the original PR have been mocked and ridiculed by the author, in original PR, yet here we are, nearly line to line, but missing the point.
This is third PR, by the author, that competes the original solution, and it seems that we are closing to getting exactly the same code in both.

> Examples of devices that do not impose write restrictions: ambiq, numaker_rmc, nrf_rram, nrf_mram and at45 in read-modify-write mode.

They have no PAGE ERASE LIMITATIONS, and at45 read-modify-write requires extra management that no common subsystem would handle (read the doc 5.3.3 Auto Page Rewrite [[Microchip, 2]](ttps://ww1.microchip.com/downloads/en/DeviceDoc/doc3443.pdf)).
You basically need to write specific solution for AT45, to be able to handle the LIMITATION OF_READ_MODIFY_WRITE.

Also Nuvoton speaks for my case [[Nuvoton, 1]](https://www.nuvoton.com/export/resource-files/en-us--PB_M2L31_Series_EN_Rev_1.00.pdf), advertising its chip as:
image
They state "no erase needed". 

> It should also not change the current understanding of the
>flash api. The PR does not try to move the flash api into a more generic storage api.
>
>The rationale behind this starting point is to provide continuity for developers that need to provide support for multiple versions of zephyr, they should not have to figure out the meaning of the flash api in different versions of zephyr.

Do I intend to move Flash API to the point where erase is removed, yes. Does it make Flash API equal to EEPROM API? Yes. Does it make both APIs redundant? Yes. And I have a good case to add erase to EEPROM, so this happens either way, wouldn't that change current understanding of EEPROM API?

The comment is really long because it takes everything that has been argued here.
 - https://discord.com/channels/720317445772017664/1220424912829681846
 - https://github.com/zephyrproject-rtos/zephyr/pull/67687
 - https://github.com/zephyrproject-rtos/zephyr/issues/64732

I will shorter your pain of going through above in this summary. If you sense my irritation, at this point, you might be correct. Prepare for a ride.
 
### Still missing the point

The original PR is intended, among other things that are realized here, to:
 1) allow subsystems to get information that instance of subsystem works on instance of a device that is NOT HAMPERED by the need to erase before write of random data a specific, previously written, address.
 2) adapt subsystem to effectively work with the capability, where something has to be done for devices that lack erase, yet code assumes that erase exists.
 3) completely fails to address issue where there are devices with single capability in the system, therefore there is no need to compile paths covering capabilities or lack thereof in subsystems.

### So once again.
 
**Addressing 1):**
Allow vs to restrict:
 - no software solution is happy to have to erase large chunks of data; software wants to change data by bits if possible; ERASING LARGE CHUNKS is HAMPERING SOFTWARE execution.
 - the software may choose to work with some capabilities of hardware, for example write-block-size == 4 may be requirement of a software imposed on hardware, here need to ERASE IS A PROBLEM IMPOSED ON SOFTWARE;
 - the ERASE is DAMAGING operation, in devices we have, that why it should be the focus, not write; all devices with ERASE HAVE LIMITED ENDURANCE, which is RESTRICTION; nearly all devices allow write but NOT ALL WRITE OPERATIONS RESTRICT ENDURANCE.

Once again what does `.unrestricted_write` mean:
 - it can be written more than once, aka not-OTP?
 - it can be written this time, because this is OTP but not yet written.
 - it can be written more than once, but device needs to be completely erased, aka UCIR?
 - it can be written more than once, but device needs to be completely erased, aka UCIR, and no write has happened yet?
 - it can be written because not read-only?
 - it can be written 1000000000000000000000 times and it does not bulge?
 - it can be written because some system does not restrict it?
 - it can be written with any value, aka RAM not Flash?
 - it can be written multiple times across write-block-size because there is no hidden ECC that requires erase?
 - it can be written with the largest possible write-block-size, because there is full power to a device? (some STM devices)
 - it can work with full speed, nothing restricts it?
 - can you really erase something you are restricted to write to? sounds like it may be read only? or OTP?
 - why entering `if (!.unrestricted_write)`, we call `flash_erase()` and not `flash_unrestrict_write()`?

Yeah, in short:
```
    if (!unrestricted_write)
        flash_erase()
    }
```

against original:
``` 
    if (explicit_erase) {
        flash_erase()
    }
```
do I have to explaing the difference?

**UPDATE 2024-03-29, 8:38**
And one more thing that I have failed to communicate: Leaving things as they are and providing the flag in a way presented here, gives false impression that everything is as it is, and user may leave the flash_erase just like it was before.
For devices that we are trying to serve here, emulation of erase brings down endurance - this needs to be communicated.
User needs to know that they have now reconsider whether they need the flash_erase at all, because their software needs that, because now their hardware may not.

If you still want to read beyond this point, I really admire that you value my efforts. You probably start to understand my irritation.

So getting back to topic: I have already stated my process on selecting the name here https://github.com/zephyrproject-rtos/zephyr/pull/67687#issuecomment-2007363746) while the naming provided within this PR, as a representation of idea, distorts the entire purpose of change.

Note also that ERASE IS MANDATORY extra step for software to be able to proceed in many cases; it is not an option that software may or may not take.
This is opposite to what this solution proposes: SOFTWARE IS MANDATED by a device to erase not allowed to not do that, which is not reflected by the change in this PR.
Would we rather have `if (extra_step_and_pain_needed)` in code or `if (!maybe_you_do_not_have_to_do_this)`.

Examples for 1):
  - it is hard to make some large erase page size devices, like STM, with MCUboot.
  - FAT driver IS PREVENTED BY ERASE from working on devices with blocks above 4k, it has no problem with working devices that have 1-4k erase size.
  - the SOFTWARE MAY CHOOSE TO WORK WITH SOME CAPABILITIES OF HARDWARE, for example write-block-size == 4 may be requirement of a software imposed on hardware, here, mostly, ERASE IS REQUIREMENT IMPOSED ON SOFTWARE.
  - Many Flash devices MAY HAVE NO RESTRICTION ON NUMBER OF WRITES per block, but MAY RESTRICT VALUE THEY CAN WRITE (only 1->0 transitions),
  - software implementation may use erase for its advantage, but it does not mean that this is considered a feature across the entire implementation.
  - erasing pages often stalls CPU, so there are special mechanisms to interrupt erase to allow code to keep executing.
  - erasing pages on external device may block other subsystems from performing read/write operations: ERASE IS HAMPERING multithreaded execution.
  - erasing pages on external device may stall XIP execution: ERASE IS HAMPERING use of external device for code.

**That is what should be represented by the flag.**

**Addressing 2):**

The solution here, conveniently, selects only one example of usage, Disk Access, where subsystem works best when it is NOT HAMPERED BY ERASE, to present that it is solving every issue, same usage that the author has considered a limited case in the original PR (https://github.com/zephyrproject-rtos/zephyr/pull/67687#issuecomment-1930591435).
The issue of original PR is not to do less work, but do enough work to make things work better. Something this PR fails to do, due to lack of understanding what erase operation is for software stack.

I have already removed MRAM and RRAM erase callbacks.

I do not find it needed for  any subsystem to check page layout or erase anything that does not require it.
Making erase call mandatory, makes it mandatory to keep pages layout for devices that have no such parameters at all.
This is keeping as back for no reason: software that just wants to erase

Also what is then against adding EEPROM devices to Flash API, if I add erase?

So again we are moving in a direction I have proposed, the difference is that you have to add extra emulation for things that devices do not have, for no reason.

Additionally, this means that instead of common subsystem code to handle situation for all instances of devices with no-erase, or completely skipping the code, THIS PR MANDATES IMPLEMENTATION OF ERASE BY DEVICES THAT DO NOT HAVE THAT HARDWARE LIMITATION!

And if somebody is willing to bring operation time table for EEPROM vs Flash [[WIKI, 4]](https://en.wikipedia.org/wiki/Flash_memory#Limitations), as argument that erase is not limitation, I would like to point out that new MRAM/RRAM devices have very similar erase/write times, to Flash, but HAVE NO ERASE LIMITATION, which EMULATED ERASE BRINGS BACK, aside from the fact that it REDUCES ENDURANCE.

No software chooses "!unrestricted_write" path in code if that is it can avoid LIMITATION OF HARDWARE; only software designed around that limitation will do that.

**Addressing 3):**

Neither Nordic devices (MRAM and RRAM based) nor Ambiq device is in Flash API for erase. They must implement it to make all subsystems work for them.

If they are sole devices in system or there are only devices of the same capability, LACK OF need to ERASE, this MEANS that these DEVICES HAVE TO CARY ERASE FUNCTIONALITY THEY DO NOT HAVE BY DESIGN.

If there is no subsystem, as a software implementation of something above driver, that does not require erase or even works efficiently without one, we still have erase implementation for no reason.

A lot of subsystems directly depend on flash, there is no way for these devices to be easily moved to EERPOM, I am explaining that in [Q&A:Issue 0](https://github.com/zephyrproject-rtos/zephyr/files/14762707/QA.no-erase-flash.pdf), I have gathered from previous discussions (still extending the PDF).

MOVING these devices, MRAM/FRAM/RRAM, under EEPROM, to be consistent, would mean MASSIVE EFFORT in DTS, subsystems, linker scripts and Kconfigs. Something we have already agreed is not acceptable when I have proposed https://github.com/zephyrproject-rtos/zephyr/issues/64732.

writing software requires understanding limitations of hardware but non-existent limitations should not hamper software execution.

No software is written for limitations, it is always written against them.

### Not a wanna-have fix, you can play with

I am addressing the issue as not only the maintainer of Flash but also an user, via being contributor or maintainer, in various other subsystems, some of them found in List of **Addressing 2):**

I have complete understanding that storage is representing one of a very few SoC subsystems that may wear SoC to death, by only performing operations it offers to an user.

I directly and indirectly represent vendors that will offer the chips with capabilities, I am trying to address, and I am quite sure they will be happy when Zephyr will use their devices in the best possible way.
I directly and indirectly represent users that will be using these chips with Zephyr.

### Flash is flash and always has erase!

This is pointless argument. Flash is EEPROM with erase separated from program [[WIKI, 3]](https://en.wikipedia.org/wiki/EEPROM#Today's_EEPROM_structure), but for the purpose of mental exercise: there is EEPROM 25aa512 [[Microchip, 1]](https://ww1.microchip.com/downloads/en/DeviceDoc/25AA256-25LC256-256K-SPI-Bus-Serial-EEPROM-20001822H.pdf) from Microchip that SUPPORTS ERASE, I have already stated my willingness https://discord.com/channels/720317445772017664/1220424912829681846/1222306705698848812 to provide driver for it that will require EEPROM API change; the change will add `.erase`, `.pages_layout` and `.parameters callback`, to support all features.
This makes the Flash and EEPROM the same API! And again, Flash is evolution of EEPROM ([[Wiki, 1]](https://en.wikipedia.org/wiki/EEPROM#), [[Wiki, 2]](https://en.wikipedia.org/wiki/EEPROM#Comparison_with_EPROM_and_EEPROM/flash), [[Wiki, 5]](https://en.wikipedia.org/wiki/Flash_memory#Background), explained in discussion https://discord.com/channels/720317445772017664/883445484923011172/1220021197064896542, https://discord.com/channels/720317445772017664/1220424912829681846/1222306705698848812, https://discord.com/channels/720317445772017664/883445484923011172/1220300256613240873).

Such a change would also lead us to a situation where you have MRAM device in Flash API, that does not provide erase callback, and at the same EEPROM device that does. The difference is that the later uses `.erase_supported` flag. The moment you add Flash that does not require erase but supports it, you have a situation where flags leak between APIs. 

There is not better case than 25AA512 that to explain that software is HAMPERED by ERASE REQUIREMENT IMPOSED ON DEVICE as supported by 25aa512 erase is great when subsystem can use the feature to pre-erase large chunk of data, for example destroy FAT FS partition, but the same FAT may work better without need to erase.

Aha: so you agree that these devices should be EEPROM? Read the **Addressing 3):** once again.

### Maintainer not a benevolent dictator

According to:
https://docs.zephyrproject.org/latest/project/project_roles.html#maintainer
as a maintainer, of Flash and other touched areas, I can put my thumb on the scale here, as I have explaining my general vision for the subsystem I maintain. I have restricted myself from doing so (note the right use of "restrict") to keep discussion open and inclusive.
The flow of the discussion around this issue hampers my ability to perform other task, as maintainer, in a timely manner and is not good for Zephyr.

Neither is this PR.

If you read to this point, please at least give me a the "eyes". Now you probably understand why anything in Flash subsystem moves so slow.

Thanks for your attention, I hope this was entertaining.

### Reference links

[WIKI, 1] Flash memory is a type of EEPROM designed for high speed and high density, at the expense of large erase block https://en.wikipedia.org/wiki/EEPROM#
[WIKI, 2] Comparison with EPROM and EEPROM/flash https://en.wikipedia.org/wiki/EEPROM#Comparison_with_EPROM_and_EEPROM/flash
[WIKI, 3] EEPROM has two transistors per bit (write/erase) Flash has one per bit one per block (write/erase) https://en.wikipedia.org/wiki/EEPROM#Today's_EEPROM_structure
[WIKI, 4] Flash, Limitations: https://en.wikipedia.org/wiki/Flash_memory#Limitations
[WIKI, 5] Flash, History: https://en.wikipedia.org/wiki/Flash_memory#Background
[Microchip, 1] 25aa512 EEPROM with support for erase, https://ww1.microchip.com/downloads/en/DeviceDoc/25AA256-25LC256-256K-SPI-Bus-Serial-EEPROM-20001822H.pdf
[Microchip, 2] AT45 https://ww1.microchip.com/downloads/en/DeviceDoc/doc3443.pdf
[Nuvoton, 1] https://www.nuvoton.com/export/resource-files/en-us--PB_M2L31_Series_EN_Rev_1.00.pdf

**Comment by Laczen** [comment]
@de-nordic, the proposed PR has been created with a starting point and a rationale behind this starting point. You can disagree with the rationale and I'm willing to discuss it. 

**The starting point of this PR is that no change should be made to any existing systems that cannot take advantage of the existence of the unrestricted write possibilities that is available. It should also not change the current understanding of the
flash api. The PR does not try to move the flash api into a more generic storage api.**

*The rationale behind this starting point is to provide continuity for developers that need to provide support for multiple versions of zephyr, they should not have to figure out the meaning of the flash api in different versions of zephyr.*

In your comment you have given quite some comments as to why this shouldn't be done implying that your proposal is blocked by this PR, while it is not. We all agree that our build system and data storage solutions have been relying on the
flash subsystem to much and this is limiting the possibilities of zephyr. It is needed to work on a more generic solution to allow zephyr to grow. What this system should look like and whether it is a device or a subsystem is not something I'm trying to solve. What I do know is that this is a major work as both the build system (using flash partitions as a means to generate the correct code location) and subsystems (upgrade possibilities, filesystems, ...) are involved. Should the storage solution provide support for files as a means for upgrade, for disk areas, ... many things need to be decided. 

The proposed PR does not limit in any way working towards this more generic storage solution. It only provides a solution for a present problem we have at hand: we need to integrate non-flash devices into zephyr that is highly relying on the flash subsystem. The PR just tries to enable the extra feature on a particular type of device as to not wear it out to fast. If this PR is only used during a limited amount of time until a more generic solution is available that is OK.

Regarding the comments about the meaning of the flag: this is just a representation of a class of devices that are governed by a specific property it is not a reflection of a physical device property. Regarding the polarity of the flag and its influence on the code: if this is such a problem the flashdisk subsystem can at init invert it.

Regarding the at45, all I did was look at the Kconfig and conclude that it was the authors intention to provide this as a "unrestricted write" device as the `READ-MODIFY-WRITE` mode was selected as default. If there is a driver error so that it can't really do this, it has to be corrected.

Regarding the need for a `flash_fill()` for a subsystem to make optimal use of the provided possibilities. The invalidation of some area might need to be done by a subsystem, how this is done is up to the subsystem, it can be done as a read-modify-write method or it can be done by a `flash_fill()`. At the moment the read-modify-write method seems like a good method but if it is deemed necessary to have a `flash_fill()` I see no problem (after evaluation) with adding this functionality to the API.

Regarding all subsystems that require change: the way this PR is applying changes does not break the current functionality of all subsystems. They might not work optimally (from the device point of view) but they work. And yes they need work to be optimized, but only after the PR gets approved.

**Comment by de-nordic** [comment]
### I do not want to take that call

The original PR now provides two safety checks, formerly it was kinda one:
 1) Need to check for flag and explicitly call the erase or alternative function - so they need to know what they are doing
 2) Lack of the erase callback will return -ENOTSUP for devices that do not implement it.

This PR lacks any safety check as it completely hides the change and makes everything look to work fine, for misaligned convenience.

If customer fails to notice this change, they may figure out too late, only when broken devices keep rolling back.

**The original PR will fail, due to the safety checks, before device is released, when device will be tested against code that the customer already has, the call to support will happen now!**
The question to the call will be "why there is no erase", and the answer will be, probably, something I have reiterated here many times... and it will be easy to explain to customer that they have to adapt, choose their path.

**The false "no-need-for-change" impression will postpone that call to the moment to the point where broken devices will start rolling back to vendors**

The call then will have two questions:
 1) Why does this happen? 
and after obvious answer, the following one will be:
 2) Why the <ugly word> your device pretends to implement callback that destroys the device?  Why this is hidden? This should return error immediately that we would know that we should adapt or call you before we release the thing! (in a very loud voice).

**I do not want to take that second type of call**

I have strong believe that none of my fellow vendor representatives and people providing professional services want to take this call either.

If I am wrong, then you are here:
image
and I wish you luck, but do not drag me to hell with you.

If that argument does not land. I have nothing more to say.

Thanks again, over and out.

**Comment by Laczen** [comment]
> ### I do not want to take that call
> The original PR now provides two safety checks, formerly it was kinda one:
> 
> 1. Need to check for flag and explicitly call the erase or alternative function - so they need to know what they are doing
> 2. Lack of the erase callback will return -ENOTSUP for devices that do not implement it.
> 
> This PR lacks any safety check as it completely hides the change and makes everything look to work fine, for misaligned convenience.
> 
> If customer fails to notice this change, they may figure out too late, only when broken devices keep rolling back.
> 
> **The original PR will fail, due to the safety checks, before device is released, when device will be tested against code that the customer already has, the call to support will happen now!** The question to the call will be "why there is no erase", and the answer will be, probably, something I have reiterated here many times... and it will be easy to explain to customer that they have to adapt, choose their path.
> 
> **The false "no-need-for-change" impression will postpone that call to the moment to the point where broken devices will start rolling back to vendors**
> 
> The call then will have two questions:
> 
> 1. Why does this happen?
>    and after obvious answer, the following one will be:
> 2. Why the ******* (some ugly word) your device pretends to implement callback that destroys the device?  Why this is hidden? This should return error immediately that we would know that we should adapt or call you before we release the thing! (in a very loud voice).
> 
> **I do not want to take that second type of call**
> 
> I have strong believe that none of my fellow vendor representatives and people providing professional services want to take this call either.
> 
> If I am wrong, then you are here: image and I wish you luck, but do not drag me to hell with you.
> 
> If that argument does not land. I have nothing more to add.
> 
> Thanks again, over and out.

It seems that you would like a warning to be issued when a subsystem is using `flash_erase()`, I have no objection to that. A `log_wrn()` or `log_wrn_once()` can be added as well as a build warning.

On what devices should this warning be issued, or should it just be added to `flash.h` and `flash_erase()`?

**Comment by de-nordic** [comment]
> > ### I do not want to take that call
> > The original PR now provides two safety checks, formerly it was kinda one:
> > 
> > 1. Need to check for flag and explicitly call the erase or alternative function - so they need to know what they are doing
> > 2. Lack of the erase callback will return -ENOTSUP for devices that do not implement it.
> > 
> > This PR lacks any safety check as it completely hides the change and makes everything look to work fine, for misaligned convenience.
> > If customer fails to notice this change, they may figure out too late, only when broken devices keep rolling back.
> > If that argument does not land. I have nothing more to add.
> > Thanks again, over and out.
> 
> It seems that you would like a warning to be issued when a subsystem is using `flash_erase()`, I have no objection to that. A `log_wrn()` or `log_wrn_once()` can be added as well as a build warning.
> 
> On what devices should this warning be issued, or should it just be added to `flash.h` and `flash_erase()`?

Let me rephrase this:
image

to: "why are you implementing something that you device does not do, it makes no sense"

There is no sense to add warning to something that is broken by design.

I find this PR broken by design.

**Comment by Laczen** [comment]
> Let me rephrase this: image
> 
> to: "why are you implementing something that you device does not do, it makes no sense, and it is neither efficient nor does good for device".
> 
> There is no sense to add warning to something that is broken by design.
> 
> I find this PR broken by design.

You can find the PR broken by design, everyone is entitled to an opinion. Would I call this PR a beauty, no. Trying to enable non-flash properties in a flash API is not what I would consider the best approach. Is expanding a flash API to non-flash devices a good alternative, no also not. But there is a need to provide a solution for these devices in our flash oriented build system and data storage.

A possible alternative is to add these devices as nvmem (or split them up if deemed necessary) and to adapt the flash_map layer into something less `flash`, add support for the nvmem devices (and possibly eeprom) and enable the use in the subsystems (including nvs and other subsystems that are directly using flash). 

Regarding the warning, don't fool yourself. If a user finds out that by doing a write to erase there is no warning or failure, but the device will fail just as fast.

**Comment by d3zd3z** [comment]
I'm going to just make a few comments here, as I've not been fully following all of the involved PRs.  As I understand, we are trying add better support for things that are somewhat flash like, but aren't exactly like flash. For example, MRAM and the likes that are more like RAM that persists.

Some things that are concerning to me are adding support for read-modify-write mode in the at45 drivers. In my mind, this is something that never should be done, because it is an operation that is never safe. There will always be a window where an untimely power loss will cause data to be loss (generally a whole page).

The thing about flash, they are weird. It is possible to abstract flash so that you get something that isn't weird. But, this is a lot of work. This is generally done by either some type of flash filesystem (littlefs), or a flash-translation layer (FTL) that makes the flash device look like a robust block device (SD cards do it this way), then with something like FAT on top of it. But of these are fairly complicated solutions, and are really hard to get right.

It is important that the driver _never_ lie about the flash device. No matter what is implemented, a filesystem or an FTL, it still needs access to a driver that honestly presents the actual capabilities of the device.

So, if we are adding devices that have more lax requirements, and really do have those more lax requirements, we need to make that very visible through the API. Flash drivers must **never** do hidden erases, and other operations, and ideally, there should be a 1:1 mapping between driver requests and the device operations. There are some things that are a bit safer, such as presenting all of the blocks as the same size, and performing multiple erases. But, this is safe because it doesn't ever lose more data than the caller would be aware of.

It would be possible to add these devices either using the flash API, but we need to extend the API so that the caller still knows what the device is capable of. An FTL or a filesystem that queries this API should still be able to use the device, and still be able to use other flash devices with restrictions.

I think part of the problem is that we don't have an FTL in Zephyr, just a filesystem. I fear people would see a driver API that is abstracting device behavior in an unsafe way, and then we would get failures in the field, because power was lost during a programming operation.

**Comment by carlescufi** [comment]
> It is important that the driver _never_ lie about the flash device. No matter what is implemented, a filesystem or an FTL, it still needs access to a driver that honestly presents the actual capabilities of the device.
> 
> So, if we are adding devices that have more lax requirements, and really do have those more lax requirements, we need to make that very visible through the API. Flash drivers must **never** do hidden erases, and other operations, and ideally, there should be a 1:1 mapping between driver requests and the device operations. There are some things that are a bit safer, such as presenting all of the blocks as the same size, and performing multiple erases. But, this is safe because it doesn't ever lose more data than the caller would be aware of.

Agreed @d3zd3z.
That is exactly the approach presented by @de-nordic in his PR https://github.com/zephyrproject-rtos/zephyr/pull/67687, via the flag whose name that is currently under debate.

**Comment by Laczen** [comment]
> It is important that the driver _never_ lie about the flash device. No matter what is implemented, a filesystem or an FTL, it still needs access to a driver that honestly presents the actual capabilities of the device.
> So, if we are adding devices that have more lax requirements, and really do have those more lax requirements, we need to make that very visible through the API. Flash drivers must **never** do hidden erases, and other operations, and ideally, there should be a 1:1 mapping between driver requests and the device operations. There are some things that are a bit safer, such as presenting all of the blocks as the same size, and performing multiple erases. But, this is safe because it doesn't ever lose more data than the caller would be aware of.

Agreed @d3zd3z, but how to get the first and second requirement matched. Honestly representing the actual capabilities of a device does not really match with adding devices that have more lax requirements. We can always "force" the devices with more lax requirements in the less lax properties, doing it the other way around is much more difficult. As an example: how does a erase-block fit into a representation of the actual capabilities of a MRAM?

Making the flash API the generalized non-volatile storage solution should at least be done using a different name. And it should be evaluated if this is the correct solution for present and future problems. At the same time it should be evaluated what it will replace, `eeprom`, `flash`, `bbram`, `retained_mem`, ...

I have come to the conclusion that neither this PR nor the PR it is an alternative for is a good proposal. Trying to add non-flash properties under a flash API should not be done. Instead focus should be moved to an alternative to `flash_map` that supports multiple types of storage.

**Comment by Laczen** [comment]
It is not needed to a property to a flash driver at all. The zephyr build system can based on devicetree definitions decide how to use a flash driver in a system above the flash drivers. Such a system is needed for partitioning anyhow. More information can be found in #70979.

**Pull request closed without merging**
